### DISCUSSION TITLE
Next.JS in a monorepo

### DISCUSSION THREAD

**Discussion by TheDSCPL** [vercel/vercel#4320] [General] [answered]
**TL;DR:**

How can I deploy a Next.JS project to Now when it's not in the root directory?

**More complete explanation:**

I am working on a monorepo and there is a Next.JS package on it under the directory `/packages/next-js`. Inside that directory there is, amongst other folders, a `src` directory and another `pages` directory inside it (so I don't have `pages` in the Next.JS root but inside `src/pages`).

Inside pages I have a `plans` directory with:

- `[...plans].tsx` - parses the plans array and displays the page accordingly
- `index.tsx` - redirects to the previous file with the default plans URL

When I run `next` **inside `packages/next-js`**, I can successfully get to `localhost:3000/plans/personal/gold` from the browser.

However, when I run `now dev`  **in the root of the monorepo** and I try to go to `localhost:3000/plans/personal/gold`, I get the 404 page. If I go to another page without a dynamic route and I click the navbar link to go to the plans, the browser displays the correct page (because it's a SPA and it just gets the right JS file due to the `next/link` component). If I refresh the page, it returns the 404 page again.

Basically the problem is configuring `now` to get the right JS file to resolve that route!

There is a `now.json` file at the monorepo's root which contains the following:

```json
{
  "name": "...",
  "version": 2,
  "scope": "...",
  "env": {
    ...
  },
  "build": {
    "env": {
      ...
    }
  },
  "builds": [
    {
      "src": "packages/next-js/package.json",
      "use": "@now/next"
    }
  ],
  "routes": [
    {
      "src": "/(.*)",
      "dest": "packages/next-js/$1",
      "headers": {
        ...
      },
      "continue": false
    }
  ],
  "public": false
}
```

I've tried prepending that catch-all route with this:
```json
    {
      "src": "/plans/.+",
      "dest": "/plans/[...plans]",
      "continue": true
    }
```
but it doesn't work and neither does
```json
    {
      "src": "/plans/(.*)",
      "dest": "packages/next-js/plans/[...plans]",
      "continue": false
    }
```

I have the `now.json` file in the root so it's possible to automatically deploy using version control.

I'm sorry but I'm getting really frustrated because I've been battling this issue for days now and there should be an easier way but the [documentation](https://vercel.com/docs/configuration#introduction/configuration-reference) is way too simplistic and doesn't cover these more advanced scenarios. Maybe if it had more technical informations such as what the `dest` _actually_ means and how it is interpreted by `now`, I'd probably be able to debug this by myself.

For instance, one of the unanswered questions is: _Why is the correct catch-all dest `"dest": "packages/next-js/$1"` and not `"dest": "packages/next-js/src/pages/$1"`_... If the one that works is the former, then the `dest` isn't actually pointing to a file. Or maybe it is but it's pointing to a file which was compiled and now lives directly under a directory tree containing `packages/next-js`. Or maybe the `@now/next` package (which doesn't appear to have its own documentation but should since it causes `now` to have a framework specific behaviour) somehow interprets the `packages/next-js/$1` and knows it's a page? I don't know. It's the "black magic" of `now`...

There should be more documentation on the deployments instead of just beginners guides and

> just type `now` and you're all set!

everywhere!

Please enlighten me on this because I don't have many more things to try and solve this issue but I need that page available from a permalink and not just by navigation.

**Comment by TheDSCPL** [comment]
Hello? Anyone?

**Reply by TheDSCPL** [reply]
@styfle @lucleray @mcsdevv Is my question just being ignored?

**Comment by samcx** [comment]
@TheDSCPL You could theoretically make monorepos work using a single `now.json` file, but we don't recommend users to go down this path.

You will be better off deploying the different projects separately (packages in your case) via zero-config or separate `now.json` files.

True monorepo support (automatically deploying multiple zero-config projects from the root) is coming in the future! Stay tuned 👀

**Reply by TheDSCPL** [reply]
Thank you very much for your answer :smiley: 
I can change the `now.json` file into the package but will it be automatically deployed on commit?

**Reply by samcx** [reply]
Monorepos won't work with Git Integration unless you configured the build settings (or `now.json` — I recommend using build settings here so you don't have to mess with writing routes) correctly. You also will be limited to one project from your monorepo.

If you want to deploy other projects in your monorepo, you will have to go to the package subdirectory and use the cli to manually deploy each project.

**Reply by TheDSCPL** [reply]
Ok! I don't have a problem with the one project per monorepo. The monorepo contains other systems written in other languages that won't be deployed to Vercel.

My question now is, how can I configure the build settings but set the headers (like CSP and HSTS) and env variables without a `now.json`?

**Reply by TheDSCPL** [reply]
I guess the true problem here is not knowing how to correctly route the requests to a project in a sub folder. If I knew how to redirect the requests while also handling dynamic routes, that would solve the problem.

**Comment by Prioe** [comment]
I've ran into a very similar problem when I was trying to setup my now/next project.

I've tried your solution

```js
  "routes": [
    {
      "src": "/(.*)",
      "dest": "packages/next-js/$1",
      "headers": {
        ...
      },
      "continue": false
    }
  ],
```

but was not able to get it working with dynamic routes.

I got it working in the end, by not using the `routes` config but `rewrites`.

I created a pull request (https://github.com/zeit/next.js/pull/12867) which adds my solution to the `with-yarn-workspaces`. This might work for you aswell.

**Reply by TheDSCPL** [reply]
Oh, I see! I thought of rewrites but then I read in the docs that `routes` was supposed to be like a shorthand version of `headers` and `rewrites` so I didn't try it! I'll give it a shot and post the result here! Thank you very much!

**Reply by TheDSCPL** [reply]
I'm using yarn workspaces so that example is applicable to me.

**Reply by TheDSCPL** [reply]
Hello, @Prioe. I still can't get to the dynamic routes even with `rewrites` and `headers`.

**Reply by Prioe** [reply]
Are you running into problems locally with `vc dev` or in a deployment? `vc dev` does have problems with `routes` and `rewrites` (https://github.com/zeit/now/issues/3777) and my PR only fixes the deployment. To run your app locally, you would use `yarn --cwd packages/next-js dev` with my approach.

If you need `rewrites` and `routes` to work locally, you might be able to create a workaround using this: https://github.com/zeit/now/issues/3777#issuecomment-594443041

**Reply by TheDSCPL** [reply]
The dynamic routes don't work both in dev and in deployment when I'm using `now`/`vercel`. In Next.js's dev, everything works, though.

**Reply by TheDSCPL** [reply]
I'm just going to put the Next package on a repo of its own and let the default routes take control. This is really annoying because if I am routing everything to the next-js package, then the Next.JS part should be receiving the dynamic routes' requests just like it does when it's in the root of the repository. Also, the rewrite's destination part isnt's exactly the path of the page, like I said in the original post. 'packages/next-js/$1` is not a valid location in the filesystem when `$1` gets replaced by the actual request path (for example `packages/next-js/foo`) because the files aren't there. So my guess is, since I'm rewriting to `packages/next-js`, it will use the resolver specified in the `builds` property, `@now/next` to handle rewrites to `packages/next-js/$1`.
If that's the case, then that `@now/next`package is correctly resolving static routes but not dynamic ones. What's really annoying is that dynamic routes don't work ONLY when there is a rewrite. They work if I don't set any configuration and keep everything in the repo's root, which is just absurd.

I'm going to try something: create a dummy project, put it in the root of a repo and use this rewrite to see if it breaks anything: "src": "/(.*)", "dest": "/$1". My guess is it won't work just because there is a rewrite.

**Comment by TheDSCPL** [comment, marked as answer]
I solved this by setting the root directory to the package's directory in the dashboard and leaving the routes untouched in the `now.json` file.
